mlflow/data.py

```diff
@@ -27,33 +33,30 @@ def _fetch_dbfs(uri, local_path):
 def _fetch_s3(uri, local_path):
     import boto3
     print("=== Downloading S3 object %s to local path %s ===" % (uri, os.path.abspath(local_path)))
-    (bucket, s3_path) = parse_s3_uri(uri)
+    (bucket, s3_path) = parse_simple_uri(uri, ["s3"])
     boto3.client('s3').download_file(bucket, s3_path, local_path)
 
 
 def _fetch_gs(uri, local_path):
     from google.cloud import storage
     print("=== Downloading GCS file %s to local path %s ===" % (uri, os.path.abspath(local_path)))
-    (bucket, gs_path) = parse_gs_uri(uri)
+    (bucket, gs_path) = parse_simple_uri(uri, ["gs"])
     storage.Client().bucket(bucket).blob(gs_path).download_to_filename(local_path)
 
 
-def parse_s3_uri(uri):
-    """Parse an S3 URI, returning (bucket, path)"""
-    parsed = urllib.parse.urlparse(uri)
-    if parsed.scheme != "s3":
-        raise Exception("Not an S3 URI: %s" % uri)
-    path = parsed.path
-    if path.startswith('/'):
-        path = path[1:]
-    return parsed.netloc, path
+def _fetch_hdfs(uri, local_path):
+    print("=== Downloading HDFS file %s to local path %s ===" % (uri, os.path.abspath(local_path)))
+    parse_simple_uri(uri, ["hdfs", "viewfs"])
+    host, port, path = _resolve_connection_params(uri)
+    with hdfs_system(host=host, port=port) as hdfs:
+        local_path = os.path.join(local_path, os.path.normpath(uri.split('/')[-1]))
```

Comment:
How does this handle URI paths like `"hdfs:/this/is/a/test/.."`? In this case `local_path` will have a trailing `/..` appended to it. Would help to add some unit tests for this

Reply:
Good catch, I made the change and added unit tests